kimchi/src/circuits/lookup/index.rs

```diff
@@ -0,0 +1,197 @@
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct LookupConstraintSystem<F: FftField> {
+    /// Lookup tables
+    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
+    pub lookup_table: Vec<DP<F>>,
+    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
+    pub lookup_table8: Vec<E<F, D<F>>>,
+
+    /// Table IDs for the lookup values.
+    /// This may be `None` if all lookups originate from table 0.
+    #[serde_as(as = "Option<o1_utils::serialization::SerdeAs>")]
+    pub table_ids: Option<DP<F>>,
+    #[serde_as(as = "Option<o1_utils::serialization::SerdeAs>")]
+    pub table_ids8: Option<E<F, D<F>>>,
+
+    /// Lookup selectors:
+    /// For each kind of lookup-pattern, we have a selector that's
+    /// 1 at the rows where that pattern should be enforced, and 0 at
+    /// all other rows.
+    #[serde_as(as = "Vec<o1_utils::serialization::SerdeAs>")]
+    pub lookup_selectors: Vec<E<F, D<F>>>,
+
+    /// Configuration for the lookup constraint.
+    #[serde(bound = "LookupConfiguration<F>: Serialize + DeserializeOwned")]
+    pub configuration: LookupConfiguration<F>,
+}
+
+impl<F: FftField + SquareRootField> LookupConstraintSystem<F> {
+    pub fn create(
+        gates: &[CircuitGate<F>],
+        lookup_tables: Vec<LookupTable<F>>,
+        domain: &EvaluationDomains<F>,
+    ) -> Result<Option<Self>, LookupError> {
+        let lookup_info = LookupInfo::<F>::create();
+
+        //~ 1. If no lookup is used in the circuit, do not create a lookup index
+        match lookup_info.lookup_used(gates) {
+            None => Ok(None),
+            Some(lookup_used) => {
+                let d1_size = domain.d1.size();
+
+                // The maximum number of entries that can be provided across all tables.
+                // Since we do not assert the lookup constraint on the final `ZK_ROWS` rows, and
+                // because the row before is used to assert that the lookup argument's final
+                // product is 1, we cannot use those rows to store any values.
+                let max_num_entries = d1_size - (ZK_ROWS as usize) - 1;
+
+                //~ 2. Get the lookup selectors and lookup tables (TODO: how?)
+                let (lookup_selectors, gate_lookup_tables) =
+                    lookup_info.selector_polynomials_and_tables(domain, gates);
+
+                //~ 3. Concatenate runtime lookup tables with the ones used by gates
+                let lookup_tables: Vec<_> = gate_lookup_tables
+                    .into_iter()
+                    .chain(lookup_tables.into_iter())
+                    .collect();
+
+                //~ 4. Get the highest number of columns `max_table_width`
+                //~    that a lookup table can have.
+                let max_table_width = lookup_tables
+                    .iter()
+                    .map(|table| table.data.len())
+                    .max()
+                    .unwrap_or(0);
+
+                //~ 5. Add the table ID stuff
```

Comment:
I'm not sure that this comment is adding value (to the code or the spec). Delete?